source/extensions/filters/network/zookeeper_proxy/utils.h

```diff
@@ -27,20 +29,53 @@ class BufferHelper : public Logger::Loggable<Logger::Id::filter> {
 public:
   BufferHelper(uint32_t max_len) : max_len_(max_len) {}
 
-  int32_t peekInt32(Buffer::Instance& buffer, uint64_t& offset);
-  int64_t peekInt64(Buffer::Instance& buffer, uint64_t& offset);
-  std::string peekString(Buffer::Instance& buffer, uint64_t& offset);
-  bool peekBool(Buffer::Instance& buffer, uint64_t& offset);
+  absl::StatusOr<int32_t> peekInt32(Buffer::Instance& buffer, uint64_t& offset);
+  absl::StatusOr<int64_t> peekInt64(Buffer::Instance& buffer, uint64_t& offset);
+  absl::StatusOr<std::string> peekString(Buffer::Instance& buffer, uint64_t& offset);
+  absl::StatusOr<bool> peekBool(Buffer::Instance& buffer, uint64_t& offset);
   void skip(uint32_t len, uint64_t& offset);
   void reset() { current_ = 0; }
 
 private:
-  void ensureMaxLen(uint32_t size);
+  absl::Status ensureMaxLen(uint32_t size);
 
   const uint32_t max_len_;
   uint32_t current_{};
 };
 
+#define ABSL_STATUS_RETURN_IF_STATUS_NOT_OK(status)                                                \
+  if (!status.ok()) {                                                                              \
+    return status;                                                                                 \
+  }
+
+#define WRITE_LOG_IF_STATUS_NOT_OK(status, log_level, ...)                                         \
+  if (!status.ok()) {                                                                              \
+    ENVOY_LOG(log_level, ##__VA_ARGS__);                                                           \
+  }
+
+#define COUNT_DECODER_ERR_AND_RETURN_IF_STATUS_NOT_OK(status)                                      \
+  if (!status.ok()) {                                                                              \
+    callbacks_.onDecodeError();                                                                    \
+    return status;                                                                                 \
+  }
+
+#define COUNT_DECODER_ERR_WITH_LOG_AND_RETURN_IF_STATUS_NOT_OK(status, log_level, ...)             \
+  if (!status.ok()) {                                                                              \
+    ENVOY_LOG(log_level, ##__VA_ARGS__);                                                           \
+    callbacks_.onDecodeError();                                                                    \
+    return status;                                                                                 \
+  }
+
+#define RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(status, message)                                   \
+  if (!status.ok()) {                                                                              \
+    return absl::InvalidArgumentError(message);                                                    \
```

Comment:
could we just return `status` then we can save a pair of macros?

Reply:
If the function return type is `absl::Status`, we could return `status` directly.
If the function return type is `absl::StatusOr<T>`, we cannot return `status` directly, so I make it return `absl::InvalidArgumentError` instead.